### PR TITLE
fix mount and umount bug

### LINKED ISSUE #2000: [BUG] failed to run sealer  alpha mount
### What happen?

sealer alpha mount my-app:v1

image

### Relevant log output?

_No response_

### What you expected to happen?

_No response_

### How to reproduce it (as minimally and precisely as possible)?

_No response_

### Anything else we need to know?

_No response_

### What is the version of Sealer you using?

sealer 0.9.1

### What is your OS environment?

_No response_

### What is the Kernel version?

_No response_

### Other environment you want to tell us?

- Cloud provider or hardware configuration:
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by Stevent-fei**
### Describe what this PR does / why we need it
fix sealer alpha mount and sealer alpha umount bug

sealer alpha mount ex:
```shell
Mount image name：
    sealer alpha mount imageName

Mount image id：
    sealer alpha mount ba15e47f5969

```

sealer alpha umount ex:
```shell
Umount image containerID：
    sealer alpha umount ba15e47f5969

Uninstall all images:
    sealer alpha umount --all
```

### Does this pull request fix one issue?

Fixes #2000

### CODE REVIEW

File: cmd/sealer/cmd/alpha/mount.go [hunk 1]
```diff
@@ -38,99 +33,104 @@ mount the cluster image to '/var/lib/sealer/data/overlay2' the directory and che
 var exampleForMountCmd = `
   sealer alpha mount(show mount list)
   sealer alpha mount my-image
+  sealer alpha mount my-image my-image1 my-image2
   sealer alpha mount ba15e47f5969
+  sealer alpha mount ba15e47f5969 69f9560200b8 8e456442ddba
 `
 
+const (
+	containerID = "CONTAINER ID"
```
Comment (Starnop) [inline cmd/sealer/cmd/alpha/mount.go:42]: add table prefix

File: cmd/sealer/cmd/alpha/mount.go [hunk 2]
```diff
@@ -38,99 +33,104 @@ mount the cluster image to '/var/lib/sealer/data/overlay2' the directory and che
...
 			images, err := store.Images()
 			if err != nil {
 				return err
 			}
 
 			//output mount list
 			if len(args) == 0 {
-				if err := mountList(images); err != nil {
-					return err
+				table := tablewriter.NewWriter(common.StdOut)
```
Comment (Starnop) [inline cmd/sealer/cmd/alpha/mount.go:68]: extract the table writer as separate method

File: cmd/sealer/cmd/alpha/mount.go (hunk 1 above)
Comment (Starnop) [reply cmd/sealer/cmd/alpha/mount.go:42]: tableHeaderContainerID

File: cmd/sealer/cmd/alpha/mount.go [hunk 3]
```diff
@@ -38,99 +33,104 @@ mount the cluster image to '/var/lib/sealer/data/overlay2' the directory and che
...
-			mounts, err := engine.Mount(&options.MountOptions{Containers: []string{cid}})
-			if err != nil {
-				return err
-			}
-
-			// remove destination dir if it exists, otherwise the Symlink will fail.
-			if _, err = os.Stat(path); err == nil {
-				return fmt.Errorf("destination directionay %s exists, you should remove it first", path)
+					for _, container := range containers {
+						if builder.ContainerID == container.ID && mounted {
```
Comment (Starnop) [inline cmd/sealer/cmd/alpha/mount.go:82]: return fast, avoid large code depth

File: cmd/sealer/cmd/alpha/umount.go [hunk 4]
```diff
@@ -50,99 +47,80 @@ func NewUmountCmd() *cobra.Command {
...
-			); err != nil {
-				return err
-			}
-
-			for _, file := range files {
-				for _, image := range images {
-					if err := removeContainerDir(image, file, imgName); err != nil {
-						return err
+				for _, container := range containers {
+					if container.ImageID == imageID {
```
Comment (Starnop) [inline cmd/sealer/cmd/alpha/umount.go:100]: if container.ImageID != imageID{
  continue
}

File: cmd/sealer/cmd/alpha/mount.go [hunk 5]
```diff
@@ -28,109 +25,181 @@ import (
...
+		return "sha256:" + id
+	}
+
+	if idTruncLength := 12; len(id) > idTruncLength {
+		return id[:idTruncLength]
+	}
+	return id
+}
+
+func contain(imageName string, images []string) bool {
```
Comment (Starnop) [inline cmd/sealer/cmd/alpha/mount.go:169]: string slice contains

File: cmd/sealer/cmd/alpha/mount.go [hunk 6]
```diff
@@ -28,109 +25,181 @@ import (
...
+					}
+				}
 			}
 
-			mountPoint := mounts[0].MountPoint
-			if err := os.Symlink(mountPoint, path); err != nil {
-				return err
+			for _, name := range args {
+				imageID := getImageID(images, name)
+				if ok := contain(imageID, imageIDList); !ok {
```
Comment (Starnop) [inline cmd/sealer/cmd/alpha/mount.go:99]: return fast

File: cmd/sealer/cmd/alpha/mount.go [hunk 7]
```diff
@@ -28,109 +25,181 @@ import (
...
+			imageID = image.ID
+		}
+		for _, name := range image.Names {
+			imgName = name
+		}
+	}
+	return imgName, imageID
+}
+
+func getImageID(images []storage.Image, name string) string {
```
Comment (Starnop) [inline cmd/sealer/cmd/alpha/mount.go:192]: how to determine the current os arch

File: cmd/sealer/cmd/alpha/mount.go (hunk 5 above)
Comment (Starnop) [reply cmd/sealer/cmd/alpha/mount.go:169]: strings.Contains

File: cmd/sealer/cmd/alpha/mount.go [hunk 8]
```diff
@@ -28,109 +25,181 @@ import (
...
+
+	if idTruncLength := 12; len(id) > idTruncLength {
+		return id[:idTruncLength]
+	}
+	return id
+}
+
+func contain(imageName string, images []string) bool {
+	for _, image := range images {
+		if imageName == image || strings.Contains(image, imageName) {
```
Comment (Starnop) [inline cmd/sealer/cmd/alpha/mount.go:171]: Contains/hasPrefix

File: cmd/sealer/cmd/alpha/mount.go [hunk 9]
```diff
@@ -16,121 +16,196 @@ package alpha
...
+	}
+	return ""
+}
 
-			mountPoint := mounts[0].MountPoint
-			if err := os.Symlink(mountPoint, path); err != nil {
-				return err
+func (m Mounter) generateMountList() error {
+	table := tablewriter.NewWriter(common.StdOut)
+	table.SetHeader([]string{imageName, tableHeaderContainerID, tableHeaderMountPath})
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/alpha/mount.go:129]: you should split table render  from this func

File: cmd/sealer/cmd/alpha/mount.go [hunk 10]
```diff
@@ -16,121 +16,196 @@ package alpha
...
+		for _, n := range image.Names {
+			if name == n {
+				return image.ID
+			}
+		}
+	}
+	return ""
+}
+
+func (m Mounter) mount(imageNameOrID []string) error {
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/alpha/mount.go:175]: Mount(imageNameOrID []string) error

File: cmd/sealer/cmd/alpha/mount.go [hunk 11]
```diff
@@ -16,121 +16,196 @@ package alpha
...
+			if builder.ContainerID == container.ID && mounted {
+				imageID := m.getMountedImageID(container)
+				imageIDList = append(imageIDList, imageID)
+			}
+		}
+	}
+
+	for _, name := range imageNameOrID {
+		imageID := m.getImageID(name)
+		if ok := utilsstrings.IsInSlice(imageID, imageIDList); !ok {
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/alpha/mount.go:192]: ok := utilsstrings.IsInSlice(imageID, imageIDList)
if ok{
logrus.Infof("this image has already been mounted, please do not repeat the operation")
}
// do others

File: cmd/sealer/cmd/alpha/umount.go [hunk 12]
```diff
@@ -15,159 +15,135 @@
...
-
-				for _, file := range files {
-					if err := os.RemoveAll(filepath.Join(common.DefaultLayerDir, file.Name())); err != nil {
-						return err
-					}
+				if err := umountInfo.umountAllContainers(); err != nil {
+					return err
 				}
-				logrus.Infof("umount all cluster image successful")
+				logrus.Infof("umount all mount information was successful")
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/alpha/umount.go:64]: successful  to umount all sealer image

File: cmd/sealer/cmd/alpha/umount.go [hunk 13]
```diff
@@ -15,159 +15,135 @@
...
-				id, err := getImageID(images, args[0])
-				if err != nil {
-					return err
-				}
-				if c.ImageID == id {
-					containerID = c.ID
-					imgName = args[0]
-				}
-			}
+func (m Mounter) umount(imageNameOrID []string) {
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/alpha/umount.go:76]: Umount(imageNameOrID []string)

File: cmd/sealer/cmd/alpha/umount.go [hunk 14]
```diff
@@ -15,159 +15,135 @@
...
-	return umountCmd
+	return ""
 }
 
-func getImageID(images []storage.Image, imageName string) (string, error) {
-	for _, image := range images {
-		for _, n := range image.Names {
-			if n == imageName {
-				return image.ID, nil
+func (m Mounter) umountAllContainers() error {
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/alpha/umount.go:119]: UmountAllContainers() error

File: cmd/sealer/cmd/alpha/umount.go [hunk 15]
```diff
@@ -15,159 +15,135 @@
...
-				for _, image := range images {
-					if err := removeContainerDir(image, file, imgName); err != nil {
-						return err
-					}
-				}
+		if err = builder.Unmount(); err != nil {
+			if lastError != nil {
+				fmt.Fprintln(os.Stderr, lastError)
 			}
+			lastError = fmt.Errorf("error unmounting container %q: %w", builder.Container, err)
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/alpha/umount.go:102]: failed to

File: cmd/sealer/cmd/alpha/umount.go [hunk 16]
```diff
@@ -15,159 +15,135 @@
...
-			if err := imageEngine.RemoveContainer(&imagecommon.RemoveContainerOptions{
-				ContainerNamesOrIDs: []string{containerID}},
-			); err != nil {
-				return err
+		builder, err := buildah.OpenBuilder(context.TODO(), m.store, containerID)
+		if err != nil {
+			if lastError != nil {
+				fmt.Fprintln(os.Stderr, lastError)
 			}
+			lastError = fmt.Errorf("error unmounting container %s: %w", name, err)
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/alpha/umount.go:91]: failed to

Comment (kakaZhou719) [inline cmd/sealer/cmd/alpha/umount.go:91]: container ?

File: cmd/sealer/cmd/alpha/umount.go [hunk 17]
```diff
@@ -15,159 +15,135 @@
...
+			containerID = name
+		} else {
+			containerID = m.getContainerID(name)
+		}
 
-			if err := imageEngine.RemoveContainer(&imagecommon.RemoveContainerOptions{
-				ContainerNamesOrIDs: []string{containerID}},
-			); err != nil {
-				return err
+		builder, err := buildah.OpenBuilder(context.TODO(), m.store, containerID)
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/alpha/umount.go:86]: what is builder? rename it to client ?

File: cmd/sealer/cmd/alpha/mount.go [hunk 18]
```diff
@@ -16,121 +16,200 @@ package alpha
...
+			}
+		}
+	}
+
+	for _, name := range imageNameOrID {
+		imageID := m.getImageID(name)
+		ok := utilsstrings.IsInSlice(imageID, imageIDList)
+		if ok {
+			logrus.Infof("this image has already been mounted, please do not repeat the operation")
+		} else {
```
Comment (Starnop) [inline cmd/sealer/cmd/alpha/mount.go:198]: continue here directly

File: cmd/sealer/cmd/alpha/umount.go [hunk 19]
```diff
@@ -15,159 +15,135 @@
...
 
-			for _, file := range files {
-				for _, image := range images {
-					if err := removeContainerDir(image, file, imgName); err != nil {
-						return err
-					}
-				}
+		if err = client.Unmount(); err != nil {
+			if lastError != nil {
+				fmt.Fprintln(os.Stderr, lastError)
```
Comment (Starnop) [inline cmd/sealer/cmd/alpha/umount.go:100]: return error here

**Review by kakaZhou719 — APPROVED** [review]
LGTM

**Pull request merged**